### PR TITLE
#1215 Remove Hardcoded Display Route for ContentItems

### PR DESCRIPTION
**Pull request by rserj**
@sebastienros in order to get right route I need to get ContentItem's Display route on every request, please review my code, thanks

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Autoroute/Routing/AutorouteRoute.cs [hunk 1]
```diff
@@ -44,31 +49,57 @@ public VirtualPathData GetVirtualPath(VirtualPathContext context)
...
+                if (displayRoutes.ContainsKey(key))
+                    context.RouteData.Values[key] = displayRoutes[key];
+            }
+            context.RouteData.Routers.Add(_target);
+        }
 
-            return Task.CompletedTask;
+        private void FillDefaultRouteData(RouteContext context, string contentItemId)
+        {
+            context.RouteData.Values["area"] = "OrchardCore.Contents";
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Autoroute/Routing/AutorouteRoute.cs]: This should be moved in a default content handler.

File: src/OrchardCore.Modules/OrchardCore.Autoroute/Routing/AutorouteRoute.cs [hunk 2]
```diff
@@ -44,31 +49,57 @@ public VirtualPathData GetVirtualPath(VirtualPathContext context)
...
+                return null;
+            var contentManager = context.RequestServices.GetService<IContentManager>();
+            var contentItem = await contentManager.GetAsync(contentItemId);
+            return contentItem == null ? null : contentManager.PopulateAspect<ContentItemMetadata>(contentItem)?.DisplayRouteValues;
+        }
+
+        private async Task EnsureRouteData(RouteContext context, string contentItemId)
+        {
+            var displayRoutes = await GetContentItemDisplayRoutes(context.HttpContext, contentItemId);
+            if (displayRoutes == null)
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Autoroute/Routing/AutorouteRoute.cs]: If the default routes are in a default handler, then just don't match anything when it's null.

File: src/OrchardCore.Modules/OrchardCore.Autoroute/Routing/AutorouteRoute.cs [hunk 3]
```diff
@@ -44,31 +49,57 @@ public VirtualPathData GetVirtualPath(VirtualPathContext context)
...
+        }
 
-                context.RouteData.Routers.Add(_target);
-                return _target.RouteAsync(context);
+        private async Task<RouteValueDictionary> GetContentItemDisplayRoutes(HttpContext context, string contentItemId)
+        {
+            if (string.IsNullOrEmpty(contentItemId))
+                return null;
+            var contentManager = context.RequestServices.GetService<IContentManager>();
+            var contentItem = await contentManager.GetAsync(contentItemId);
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Autoroute/Routing/AutorouteRoute.cs]: Can you debug the code to ensure that when the default action is hit then it already finds the content item from the cache and doesn't reload it from database?

Then we need to confirm there is no performance impact on both displaying a content item and rendering links.

File: src/OrchardCore.Modules/OrchardCore.Autoroute/Routing/AutorouteRoute.cs [hunk 4]
```diff
@@ -21,21 +24,23 @@ public AutorouteRoute(IAutorouteEntries entries, IRouter target)
...
         {
-            if (context.Values["area"]?.ToString() == "OrchardCore.Contents" &&
-                context.Values["controller"]?.ToString() == "Item" &&
-                context.Values["action"]?.ToString() == "Display" &&
-                context.Values["contentItemId"] != null)
-            {
-                // Matching route value, lookup for the id
+            string contentItemId = context.Values["contentItemId"]?.ToString();
+            
+            if (string.IsNullOrEmpty(contentItemId))
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Autoroute/Routing/AutorouteRoute.cs]: Formatting: we always put braces.

File: src/OrchardCore.Modules/OrchardCore.Autoroute/Routing/AutorouteRoute.cs (hunk 3 above)
Comment (rserj) [reply src/OrchardCore.Modules/OrchardCore.Autoroute/Routing/AutorouteRoute.cs]: Sure, I checked in SQL Profiler

exec sp_executesql N'select  TOP 1 [Document].* from [Document] inner join [ContentItemIndex] on [ContentItemIndex].[DocumentId] = [Document].[Id]  where (([ContentItemIndex].[ContentItemId] = @p0) and ([ContentItemIndex].[Published] = @p1))',N'@p0 nvarchar(4000),@p1 bit',@p0=N'4by24ckrbhcx16d4p988brck4w',@p1=1


Database got hit only once in AutorouteRoute, then in ItemController ContentItem returned from YesSql cache

File: src/OrchardCore.Modules/OrchardCore.Autoroute/Routing/AutorouteRoute.cs (hunk 2 above)
Comment (rserj) [reply src/OrchardCore.Modules/OrchardCore.Autoroute/Routing/AutorouteRoute.cs]: fixed

**Pull request merged**